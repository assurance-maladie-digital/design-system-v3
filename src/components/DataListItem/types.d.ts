import { useCustomizableOptions } from '@/composables/customizable'
import { IndexedObject } from '@/types'

export interface DataListItem {
	key: string
	value?: string | number
	action?: string
	chip?: boolean
	icon?: string
	options?: useCustomizableOptions
	class?: string
}

export type DataList = DataListItem[]

export interface DataListIcons {
	[iconName: string]: string
}

export type ItemClass = (string | undefined | IndexedObject<boolean>)[]
