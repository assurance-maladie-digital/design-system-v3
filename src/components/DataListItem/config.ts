import type { VariantType } from '@/types/vuetifyTypes'

export const config = {
	icon: {
		size: 24,
		class: 'mr-4 mt-2',
	},
	chip: {
		class: 'mt-1',
	},
	actionBtn: {
		variant: 'text' as VariantType,
		size: 'small',
		color: 'secondary',
		class: 'text-body-1 pa-0',
	},
}
