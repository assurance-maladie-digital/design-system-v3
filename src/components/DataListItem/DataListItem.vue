<script lang="ts" setup>
	import { computed } from 'vue'
	import { useTheme } from 'vuetify'

	import useCustomizableOptions, { type CustomizableOptions } from '@/composables/useCustomizableOptions'

	import { config } from './config'
	import { locales } from './locales'

	const props = withDefaults(defineProps<CustomizableOptions & {
		label: string
		value?: string | number
		action?: string
		placeholder?: string
		chip?: boolean
		icon?: string
		row?: boolean
		renderHtmlValue?: boolean
	}>(), {
		label: '',
		value: undefined,
		action: undefined,
		placeholder: locales.placeholder,
		chip: false,
		icon: undefined,
		row: false,
		renderHtmlValue: false,
	})

	const emits = defineEmits(['click:action'])

	const options = useCustomizableOptions(config, props)
	const theme = useTheme()

	const labelColor = computed(() => {
		return theme.current.value.dark
			? 'rgba(255, 255, 255, .7)'
			: 'rgba(0, 0, 0, .6)'
	})

	const itemValue = computed(() => {
		if (typeof props.value === 'number') {
			return isNaN(props.value)
				? props.placeholder
				: props.value.toString()
		}
		return props.value || props.placeholder
	})
</script>

<template>
	<li class="vd-data-list-item d-flex flex-wrap">
		<slot name="icon">
			<VIcon
				v-if="icon"
				v-bind="options.icon"
			>
				{{ icon }}
			</VIcon>
		</slot>

		<div class="vd-data-list-item-content">
			<div :class="{ 'vd-row': row }">
				<div
					:style="{ color: labelColor }"
					class="vd-data-list-item-label text-caption"
				>
					{{ label }}
				</div>

				<div class="vd-data-list-item-value">
					<slot
						name="value"
						v-bind="{ itemValue }"
					>
						<VChip
							v-if="chip"
							v-bind="options.chip"
						>
							{{ itemValue }}
						</VChip>

						<span
							v-else-if="renderHtmlValue"
							class="text-body-1"
							v-html="itemValue"
						/>

						<span
							v-else
							class="text-body-1"
							v-text="itemValue"
						/>
					</slot>
				</div>
			</div>

			<slot name="action">
				<VBtn
					v-if="action"
					class="vd-data-list-item-action-btn"
					v-bind="options.actionBtn"
					@click="emits('click:action')"
				>
					{{ action }}
				</VBtn>
			</slot>
		</div>
	</li>
</template>

<style lang="scss" scoped>
.vd-row {
  display: flex;
  flex-wrap: wrap;

  .vd-data-list-item-label {
    align-self: center;

    &::after {
      content: ':';
      margin: 0 4px;
    }
  }
}

.vd-data-list-item-action-btn.v-btn {
  min-width: 0;
  margin: 0 -1px;
}

.v-icon.v-theme--light {
  color: rgba(0, 0, 0, 0.54);
}
</style>
