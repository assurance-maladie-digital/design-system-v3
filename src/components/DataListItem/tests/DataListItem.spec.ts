import { mount } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'
import { vuetify } from '@tests/unit/setup'

import DataListItem from '../DataListItem.vue'

describe('DataListItem', () => {
	it('renders correctly', () => {
		const wrapper = mount(DataListItem, {
			global: {
				plugins: [vuetify],
			},
			props: {
				label: 'Test',
				value: 'Sample Value',
				action: 'Click Me',
				chip: false, // Set props based on your needs
				row: true,
			},
			slots: {
				icon: '<VIcon>mdi-home</VIcon>',
				value: '<span>Custom Value</span>',
				action: '<VBtn>Custom Action</VBtn>',
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly with a number value', () => {
		const wrapper = mount(DataListItem, {
			global: {
				plugins: [vuetify],
			},
			props: {
				label: 'Test',
				value: 123,
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly with a NaN value', () => {
		const wrapper = mount(DataListItem, {
			global: {
				plugins: [vuetify],
			},
			props: {
				label: 'Test',
				value: parseInt('test', 10),
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly a value with HTML as text', () => {
		const wrapper = mount(DataListItem, {
			global: {
				plugins: [vuetify],
			},
			props: {
				label: 'Test',
				value: '<span>Custom Value</span>',
			},
		})

		const elValue = wrapper.find('.vd-data-list-item-value span')

		expect(elValue.text()).toBe('<span>Custom Value</span>')
		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly a value as plain HTML', () => {
		const wrapper = mount(DataListItem, {
			global: {
				plugins: [vuetify],
			},
			props: {
				label: 'Test',
				value: 'Paul<br> Dupont',
				renderHtmlValue: true,
			},
		})

		const elValue = wrapper.find('.vd-data-list-item-value span')

		expect(elValue.text()).toBe('Paul Dupont')
		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly value in a chip', () => {
		const wrapper = mount(DataListItem, {
			global: {
				plugins: [vuetify],
			},
			props: {
				label: 'Test',
				value: 'Chip Value',
				chip: true,
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly with an action', () => {
		const wrapper = mount(DataListItem, {
			global: {
				plugins: [vuetify],
			},
			props: {
				label: 'Test',
				action: 'Click Me',
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('emits click:action event on button click', async () => {
		const wrapper = mount(DataListItem, {
			global: {
				plugins: [vuetify],
			},
			props: {
				label: 'Test',
				action: 'Click Me',
			},
		})

		const button = wrapper.find('.vd-data-list-item-action-btn')
		await button.trigger('click')

		expect(wrapper.emitted('click:action')).toBeTruthy()
	})

	it('renders correctly in row mode', () => {
		const wrapper = mount(DataListItem, {
			global: {
				plugins: [vuetify],
			},
			props: {
				label: 'Test',
				value: 'Sample Value',
				action: 'Click Me',
				chip: false,
				row: true,
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})
})
