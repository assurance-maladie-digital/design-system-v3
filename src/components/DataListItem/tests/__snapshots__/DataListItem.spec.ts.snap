// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataListItem > renders correctly 1`] = `
"<li data-v-e715a895="" class="vd-data-list-item d-flex flex-wrap"><i class="mdi-home mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i>
  <div data-v-e715a895="" class="vd-data-list-item-content">
    <div data-v-e715a895="" class="vd-row">
      <div data-v-e715a895="" class="vd-data-list-item-label text-caption" style="color: rgba(0, 0, 0, .6);">Test</div>
      <div data-v-e715a895="" class="vd-data-list-item-value"><span>Custom Value</span></div>
    </div><button type="button" class="v-btn v-btn--elevated v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator="">Custom Action</span>
      <!---->
      <!---->
    </button>
  </div>
</li>"
`;

exports[`DataListItem > renders correctly a value as plain HTML 1`] = `
"<li data-v-e715a895="" class="vd-data-list-item d-flex flex-wrap">
  <!--v-if-->
  <div data-v-e715a895="" class="vd-data-list-item-content">
    <div data-v-e715a895="" class="">
      <div data-v-e715a895="" class="vd-data-list-item-label text-caption" style="color: rgba(0, 0, 0, .6);">Test</div>
      <div data-v-e715a895="" class="vd-data-list-item-value"><span data-v-e715a895="" class="text-body-1">Paul<br> Dupont</span></div>
    </div>
    <!--v-if-->
  </div>
</li>"
`;

exports[`DataListItem > renders correctly a value with HTML as text 1`] = `
"<li data-v-e715a895="" class="vd-data-list-item d-flex flex-wrap">
  <!--v-if-->
  <div data-v-e715a895="" class="vd-data-list-item-content">
    <div data-v-e715a895="" class="">
      <div data-v-e715a895="" class="vd-data-list-item-label text-caption" style="color: rgba(0, 0, 0, .6);">Test</div>
      <div data-v-e715a895="" class="vd-data-list-item-value"><span data-v-e715a895="" class="text-body-1">&lt;span&gt;Custom Value&lt;/span&gt;</span></div>
    </div>
    <!--v-if-->
  </div>
</li>"
`;

exports[`DataListItem > renders correctly in row mode 1`] = `
"<li data-v-e715a895="" class="vd-data-list-item d-flex flex-wrap">
  <!--v-if-->
  <div data-v-e715a895="" class="vd-data-list-item-content">
    <div data-v-e715a895="" class="vd-row">
      <div data-v-e715a895="" class="vd-data-list-item-label text-caption" style="color: rgba(0, 0, 0, .6);">Test</div>
      <div data-v-e715a895="" class="vd-data-list-item-value"><span data-v-e715a895="" class="text-body-1">Sample Value</span></div>
    </div><button data-v-e715a895="" type="button" class="v-btn v-theme--light text-accent v-btn--density-default v-btn--size-small v-btn--variant-text text-body-1 pa-0 vd-data-list-item-action-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator="">Click Me</span>
      <!---->
      <!---->
    </button>
  </div>
</li>"
`;

exports[`DataListItem > renders correctly value in a chip 1`] = `
"<li data-v-e715a895="" class="vd-data-list-item d-flex flex-wrap">
  <!--v-if-->
  <div data-v-e715a895="" class="vd-data-list-item-content">
    <div data-v-e715a895="" class="">
      <div data-v-e715a895="" class="vd-data-list-item-label text-caption" style="color: rgba(0, 0, 0, .6);">Test</div>
      <div data-v-e715a895="" class="vd-data-list-item-value"><span data-v-e715a895="" class="v-chip v-theme--light v-chip--density-default v-chip--size-default v-chip--variant-tonal mt-1" draggable="false"><!----><span class="v-chip__underlay"></span>
        <!---->
        <!---->
        <div class="v-chip__content" data-no-activator="">Chip Value</div>
        <!---->
        <!----></span>
      </div>
    </div>
    <!--v-if-->
  </div>
</li>"
`;

exports[`DataListItem > renders correctly with a NaN value 1`] = `
"<li data-v-e715a895="" class="vd-data-list-item d-flex flex-wrap">
  <!--v-if-->
  <div data-v-e715a895="" class="vd-data-list-item-content">
    <div data-v-e715a895="" class="">
      <div data-v-e715a895="" class="vd-data-list-item-label text-caption" style="color: rgba(0, 0, 0, .6);">Test</div>
      <div data-v-e715a895="" class="vd-data-list-item-value"><span data-v-e715a895="" class="text-body-1">…</span></div>
    </div>
    <!--v-if-->
  </div>
</li>"
`;

exports[`DataListItem > renders correctly with a number value 1`] = `
"<li data-v-e715a895="" class="vd-data-list-item d-flex flex-wrap">
  <!--v-if-->
  <div data-v-e715a895="" class="vd-data-list-item-content">
    <div data-v-e715a895="" class="">
      <div data-v-e715a895="" class="vd-data-list-item-label text-caption" style="color: rgba(0, 0, 0, .6);">Test</div>
      <div data-v-e715a895="" class="vd-data-list-item-value"><span data-v-e715a895="" class="text-body-1">123</span></div>
    </div>
    <!--v-if-->
  </div>
</li>"
`;

exports[`DataListItem > renders correctly with an action 1`] = `
"<li data-v-e715a895="" class="vd-data-list-item d-flex flex-wrap">
  <!--v-if-->
  <div data-v-e715a895="" class="vd-data-list-item-content">
    <div data-v-e715a895="" class="">
      <div data-v-e715a895="" class="vd-data-list-item-label text-caption" style="color: rgba(0, 0, 0, .6);">Test</div>
      <div data-v-e715a895="" class="vd-data-list-item-value"><span data-v-e715a895="" class="text-body-1">…</span></div>
    </div><button data-v-e715a895="" type="button" class="v-btn v-theme--light text-accent v-btn--density-default v-btn--size-small v-btn--variant-text text-body-1 pa-0 vd-data-list-item-action-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator="">Click Me</span>
      <!---->
      <!---->
    </button>
  </div>
</li>"
`;
